### PR TITLE
Throw unimplemented exception for complex calldata arrays.

### PR DESCRIPTION
**Pull request by chriseth**
Required for https://github.com/ethereum/solidity/pull/4461

### CODE REVIEW

File: libsolidity/codegen/ABIFunctions.cpp [hunk 1]
```diff
@@ -1188,7 +1188,8 @@ string ABIFunctions::abiDecodingFunctionCalldataArray(ArrayType const& _type)
 	solAssert(_type.dataStoredIn(DataLocation::CallData), "");
 	if (!_type.isDynamicallySized())
 		solAssert(_type.length() < u256("0xffffffffffffffff"), "");
-	solAssert(!_type.baseType()->isDynamicallyEncoded(), "");
+	if (_type.baseType()->isDynamicallyEncoded())
+		solUnimplemented("Calldata arrays with non-value base types are not yet supported by Solidity.");
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.cpp:1192]: Not sure whether ``by Solidity`` is actually needed, but in general the change looks fine.

**Review by ekpyron — APPROVED** [review]

File: libsolidity/codegen/ABIFunctions.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp:1192]: I wanted to make the distinction between "not supported by ABIEncoderV2" or "by Solidity as a whole".

**Comment by axic** [comment]
Can we add a proper typechecker check for this?

**Comment by chriseth** [comment]
Created an issue: https://github.com/ethereum/solidity/issues/4490

**Pull request merged**